process/sync/basicForkDetector.go

```diff
@@ -71,55 +82,58 @@ func (bfd *basicForkDetector) AddHeader(header data.HeaderHandler, hash []byte,
 
 	if state == process.BHProcessed {
 		// create a check point and remove all the past headers
-		bfd.mutFork.Lock()
-		bfd.fork.lastCheckpointNonce = bfd.fork.checkpointNonce
-		bfd.fork.lastCheckpointRound = bfd.fork.checkpointRound
-		bfd.fork.checkpointNonce = header.GetNonce()
-		bfd.fork.checkpointRound = int64(header.GetRound())
-		bfd.mutFork.Unlock()
+		if finalHeader == nil || finalHeader.IsInterfaceNil() {
+			bfd.setFinalCheckpoint(bfd.lastCheckpoint())
+		} else if finalHeader.GetNonce() > core.GenesisBlockNonce {
+			bfd.setFinalCheckpoint(&checkpointInfo{nonce: finalHeader.GetNonce(), round: finalHeader.GetRound()})
+			bfd.append(&headerInfo{
+				nonce: finalHeader.GetNonce(),
+				round: finalHeader.GetRound(),
+				hash:  finalHeaderHash,
+				state: process.BHNotarized,
+			})
+		}
+
+		bfd.addCheckpoint(&checkpointInfo{nonce: header.GetNonce(), round: header.GetRound()})
+
 		bfd.removePastHeaders()
 		bfd.removeInvalidHeaders()
+		bfd.removePastCheckpoints()
 	}
 
 	bfd.append(&headerInfo{
 		nonce: header.GetNonce(),
 		round: header.GetRound(),
-		hash:  hash,
+		hash:  headerHash,
 		state: state,
 	})
 
 	probableHighestNonce := bfd.computeProbableHighestNonce()
-
-	bfd.mutFork.RLock()
-	bfd.fork.lastBlockRound = bfd.rounder.Index()
-	bfd.fork.probableHighestNonce = probableHighestNonce
-	bfd.mutFork.RUnlock()
+	bfd.setLastBlockRound(uint64(bfd.rounder.Index()))
+	bfd.setProbableHighestNonce(probableHighestNonce)
 
 	return nil
 }
 
 func (bfd *basicForkDetector) checkBlockValidity(header data.HeaderHandler, state process.BlockHeaderState) error {
-	bfd.mutFork.RLock()
-	roundDif := int64(header.GetRound()) - bfd.fork.lastCheckpointRound
-	nonceDif := int64(header.GetNonce() - bfd.fork.lastCheckpointNonce)
-	bfd.mutFork.RUnlock()
+	roundDif := int64(header.GetRound()) - int64(bfd.finalCheckpoint().round)
+	nonceDif := int64(header.GetNonce()) - int64(bfd.finalCheckpoint().nonce)
+	nextRound := bfd.rounder.Index() + 1
 
 	if roundDif <= 0 {
 		return ErrLowerRoundInBlock
 	}
 	if nonceDif <= 0 {
 		return ErrLowerNonceInBlock
 	}
-	if int64(header.GetRound()) > bfd.rounder.Index() {
+	if int64(header.GetRound()) > nextRound {
 		return ErrHigherRoundInBlock
 	}
-	if header.GetNonce() == bfd.fork.checkpointNonce {
-		roundTooOld := int64(header.GetRound()) < bfd.rounder.Index()-process.ForkBlockFinality
-		if roundTooOld {
-			return ErrLowerRoundInBlock
-		}
+	roundTooOld := int64(header.GetRound()) < bfd.rounder.Index()-process.ForkBlockFinality
```

Comment:
move roundTooOld calculation to the top of the function as roundDif and nonceDif

Reply:
Done